edb/pgsql/types.py

```diff
@@ -384,16 +384,51 @@ def get_pointer_storage_info(
 
 class PointerStorageInfo(NamedTuple):
 
-    table_name: Tuple[str, str]
+    table_name: Optional[Tuple[str, str]]
     table_type: str
     column_name: str
     column_type: Tuple[str, str]
 
 
-@functools.lru_cache()
+@overload
 def get_ptrref_storage_info(
-        ptrref: irast.PointerRef, *,
-        source=None, resolve_type=True, link_bias=False):
+    ptrref: irast.BasePointerRef, *,
+    resolve_type: bool=...,
+    link_bias: Literal[False]=False,
+    allow_missing: Literal[False]=False,
+) -> PointerStorageInfo:
+    ...
+
+
+@overload
+def get_ptrref_storage_info(  # NoQA: F811
+    ptrref: irast.BasePointerRef, *,
+    resolve_type: bool=...,
+    link_bias: bool=...,
+    allow_missing: bool=...,
+) -> Optional[PointerStorageInfo]:
+    ...
+
+
+def get_ptrref_storage_info(  # NoQA: F811
+        ptrref: irast.BasePointerRef, *,
+        resolve_type=True, link_bias=False,
+        allow_missing=False) -> Optional[PointerStorageInfo]:
+    # We wrap the real version because of bad mypy interactions
+    # with lru_cache.
+    return _get_ptrref_storage_info(
+        ptrref,
+        resolve_type=resolve_type,
+        link_bias=link_bias,
+        allow_missing=allow_missing,
+    )
+
+
+@functools.lru_cache()
```

Comment:
Good call on adding the cache here.

Reply:
There was already a cache, I just needed to shuffle around how it works.